art/attacks/inference/membership_inference.py

```diff
@@ -0,0 +1,346 @@
+
+"""
+This module implements membership inference attacks.
+
+"""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+from typing import Optional
+
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.optim as optim
+from torch.utils.data.dataset import Dataset
+from torch.utils.data import DataLoader
+
+from art.attacks import InferenceAttack
+from art.estimators.estimator import BaseEstimator, NeuralNetworkMixin
+from art.estimators.classification.classifier import ClassifierMixin, Classifier
+from art.utils import check_and_transform_label_format
+from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
+
+logger = logging.getLogger(__name__)
+
+use_cuda = torch.cuda.is_available()
+
+
+def to_cuda(x):
+    if use_cuda:
+        x = x.cuda()
+    return x
+
+
+class AttackDataset(Dataset):
+    def __init__(self, x1, x2, y=None):
+        self.x1 = torch.from_numpy(x1.astype(np.float64)).type(torch.FloatTensor)
+        self.x2 = torch.from_numpy(x2.astype(np.int32)).type(torch.FloatTensor)
+
+        if y is not None:
+            self.y = torch.from_numpy(y.astype(np.int8)).type(torch.FloatTensor)
+        else:
+            self.y = torch.zeros(x1.shape[0])
+
+    def __len__(self):
+        return len(self.x1)
+
+    def __getitem__(self, idx):
+        if idx >= len(self.x1):
+            raise IndexError("Invalid Index")
+
+        return self.x1[idx], self.x2[idx], self.y[idx]
+
+
+class MembershipInferenceAttackModel(nn.Module):
+
+    def __init__(self, num_classes, num_features=None):
+
+        self.num_classes = num_classes
+        if num_features:
+            self.num_features = num_features
+        else:
+            self.num_features = num_classes
+
+        super(MembershipInferenceAttackModel, self).__init__()
+
+        self.features = nn.Sequential(
+            nn.Linear(self.num_features, 512),
+            nn.ReLU(),
+            nn.Linear(512, 100),
+            nn.ReLU(),
+            nn.Linear(100, 64),
+            nn.ReLU(),
+        )
+
+        self.labels = nn.Sequential(
+            nn.Linear(self.num_classes, 256),
+            nn.ReLU(),
+            nn.Linear(256, 64),
+            nn.ReLU(),
+        )
+
+        self.combine = nn.Sequential(
+            nn.Linear(64 * 2, 1),
+        )
+
+        self.output = nn.Sigmoid()
+
+    def forward(self, x1, l):
+        out_x1 = self.features(x1)
+        out_l = self.labels(l)
+        is_member = self.combine(torch.cat((out_x1, out_l), 1))
+        return self.output(is_member)
+
+
+class MembershipInferenceBlackBoxRuleBased(InferenceAttack):
+    """
+        Implementation of a simple, rule-based black-box membership inference attack.
+
+        This implementation uses the simple rule: if the model's prediction for a sample is correct, then it is a
+        member. Otherwise, it is not a member.
+    """
+    _estimator_requirements = [BaseEstimator]
```

Comment:
Does `MembershipInferenceBlackBoxRuleBased` require classification models? If yes, we should add `ClassifierMixin` to `_estimator_requirements`.